crates/bevy_ui/src/geometry.rs

```diff
@@ -45,6 +55,70 @@ pub enum Val {
     VMax(f32),
 }
 
+#[derive(Debug, Error, PartialEq, Eq)]
+pub enum ValParseError {
+    UnitMissing,
+    ValueMissing,
+    InvalidValue,
+    InvalidUnit,
+}
+
+impl core::fmt::Display for ValParseError {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        match self {
+            ValParseError::UnitMissing => write!(f, "unit missing"),
+            ValParseError::ValueMissing => write!(f, "value missing"),
+            ValParseError::InvalidValue => write!(f, "invalid value"),
+            ValParseError::InvalidUnit => write!(f, "invalid unit"),
+        }
+    }
+}
+
+impl core::str::FromStr for Val {
+    type Err = ValParseError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let s = s.trim();
+
+        if s.eq_ignore_ascii_case("auto") {
+            return Ok(Val::Auto);
+        }
+
+        let Some(end_of_number) = s
```

Comment:
Only small nit with this PR is that this doesn't accommodate all of rust's valid `f32` parsing inputs; I don't think people will often want to make values of `1e5 px` or whatever, but it's possible that values like that may show up in automatically-formatted float output, for example.

Reply:
This doesn't support: `NaN`, `inf`, `infinity`, `+inf`, `1E4` and `1e5`. I think most developers won't need this, and it can be added later

Reply:
Indeed. That's why I approved it :)